pkg/providers/snow/defaults.go

```diff
@@ -25,49 +26,49 @@ func NewDefaulters(aws aws.Clients, writer filewriter.FileWriter) *Defaulters {
 func NewDefaultersFromAwsClientMap(awsClientMap AwsClientMap, writer filewriter.FileWriter) *Defaulters {
 	return &Defaulters{
 		awsClientMap: awsClientMap,
+		writer:       writer,
 	}
 }
 
-func (d *Defaulters) GenerateDefaultSshKey(ctx context.Context, m *v1alpha1.SnowMachineConfig) (keyName, keyVal string, err error) {
-	if m.Spec.SshKeyName != "" {
-		return "", "", nil
-	}
-
-	keyName = fmt.Sprintf("%s-%s.pem", defaultAwsSshKeyNamePrefix, m.GetName())
+func (d *Defaulters) GenerateDefaultSshKeys(ctx context.Context, machineConfigs map[string]*v1alpha1.SnowMachineConfig) error {
+	keyGenerated := false
 
-	logger.V(1).Info("SnowMachineConfig SshKey is empty. Creating default key pair", "default key name", keyName)
-
-	// Only need to fetch IP from single device to create ssh key pair. The created key will then be validated in validateSshKeyPair method
-	var anyDeviceIP string
-	for ip := range d.awsClientMap {
-		anyDeviceIP = ip
-		break
-	}
-	keyVal, err = d.awsClientMap[anyDeviceIP].EC2CreateKeyPair(ctx, keyName)
-	if err != nil {
-		return "", "", fmt.Errorf("creating ssh key pair in snow: %v", err)
+	for _, m := range machineConfigs {
+		if err := d.SetupDefaultSshKey(ctx, m, &keyGenerated); err != nil {
+			return err
+		}
 	}
 
-	return keyName, keyVal, nil
+	return nil
 }
 
-func (d *Defaulters) GenerateDefaultSshKeyAndSaveToFile(ctx context.Context, m *v1alpha1.SnowMachineConfig) error {
+func (d *Defaulters) SetupDefaultSshKey(ctx context.Context, m *v1alpha1.SnowMachineConfig, keyGenerated *bool) error {
```

Comment:
I would make this method part of a different type
Like `MachineConfigsDefaulter`. And that type can contain the boolean
Just to avoid passing a flag as an argument :)